blackjax/mcmc/ghmc.py

```diff
@@ -0,0 +1,209 @@
+"""Public API for the Generalized (Non-reversible w/ persistent momentum) HMC Kernel"""
+from typing import Callable, NamedTuple, Optional, Tuple
+
+import jax
+import jax.numpy as jnp
+from jax.flatten_util import ravel_pytree
+
+import blackjax.mcmc.integrators as integrators
+import blackjax.mcmc.proposal as proposal
+from blackjax.types import PRNGKey, PyTree
+
+
+class GHMCState(NamedTuple):
+    position: PyTree
+    momentum: PyTree
+    potential_energy: float
+    potential_energy_grad: PyTree
+    slice: float
+
+
+class GHMCInfo(NamedTuple):
+    momentum: PyTree
+    acceptance_probability: float
+    is_accepted: bool
+    is_divergent: bool
+    energy: float
+    proposal: proposal.Proposal
+
+
+def init(
+    rng_key: PRNGKey,
+    position: PyTree,
+    logprob_fn: Callable,
+    logprob_grad_fn: Optional[Callable] = None,
+):
+    def potential_fn(x):
+        return -logprob_fn(x)
+
+    if logprob_grad_fn:
+        potential_energy_grad = jax.tree_map(
+            lambda g: -1.0 * g, logprob_grad_fn(position)
+        )
+        potential_energy = potential_fn(position)
+
+    else:
+        potential_energy, potential_energy_grad = jax.value_and_grad(potential_fn)(
+            position
+        )
+
+    p, unravel_fn = ravel_pytree(position)
+    key_mometum, key_slice = jax.random.split(rng_key)
+    momentum = unravel_fn(jax.random.normal(key_mometum, p.shape))
+    slice = jax.random.uniform(key_slice, minval=-1.0, maxval=1.0)
+
+    return GHMCState(position, momentum, potential_energy, potential_energy_grad, slice)
+
+
+def kernel(
+    noise_gn: Callable = lambda _: 0.0,
+    divergence_threshold: float = 1000,
+):
+    def kinetic_energy_fn(momentum):
+        momentum, _ = ravel_pytree(momentum)
+        return 0.5 * jnp.dot(momentum, momentum)
+
+    def one_step(
+        rng_key: PRNGKey,
+        state: GHMCState,
+        logprob_fn: Callable,
+        step_size: PyTree,  # float,
+        alpha: float,
+        delta: float,
+        logprob_grad_fn: Optional[Callable] = None,
+    ):
+        def potential_fn(x):
+            return -logprob_fn(x)
+
+        proposal_generator = persistent_ghmc_proposal(
+            potential_fn,
+            kinetic_energy_fn,
+            step_size,
+            divergence_threshold,
+            logprob_grad_fn,
+        )
+
+        key_momentum, key_noise = jax.random.split(rng_key)
+        position, momentum, potential_energy, potential_energy_grad, slice = state
+        momentum, unravel_fn = ravel_pytree(momentum)
+        # New momentum is persistent
+        momentum = unravel_fn(
```

Comment:
We can re-use the code in `metrics.py` for everything momentum related, no?

Reply:
The code now uses `metrics.py` (I've also commented a version without `metrics.py`) but it seems unnecessary. Maybe I could write a version of `gaussian_euclidean` with persistent momentum?